tools/slang-fiddle/slang-fiddle-template.cpp

```diff
@@ -0,0 +1,530 @@
+
+            case ' ':
+            case '\t':
+                lineCursor++;
+                continue;
+
+            case '%':
+                return lineCursor;
+            }
+        }
+        return nullptr;
+    }
+
+    List<RefPtr<TextTemplateStmt>> stmts;
+
+    void addRaw(char const* rawBegin, char const* rawEnd)
+    {
+        if (rawBegin == rawEnd)
+            return;
+
+        auto stmt = RefPtr(new TextTemplateRawStmt());
+        stmt->text = UnownedStringSlice(rawBegin, rawEnd);
+        stmts.add(stmt);
+    }
+
+    void addScriptStmtLine(char const* sourceBegin, char const* sourceEnd)
+    {
+        auto stmt = RefPtr(new TextTemplateScriptStmt());
+        stmt->scriptSource = UnownedStringSlice(sourceBegin, sourceEnd);
+        stmts.add(stmt);
+    }
+
+    void addScriptSpliceExpr(char const* sourceBegin, char const* sourceEnd)
+    {
+        auto stmt = RefPtr(new TextTemplateSpliceStmt());
+        stmt->scriptExprSource = UnownedStringSlice(sourceBegin, sourceEnd);
+        stmts.add(stmt);
+    }
+
+    bool isIdentifierStartChar(int c)
+    {
+        return (('a' <= c) && (c <= 'z')) || (('A' <= c) && (c <= 'Z')) || (c == '_');
+    }
+
+    bool isIdentifierChar(int c) { return isIdentifierStartChar(c) || (('0' <= c) && (c <= '9')); }
+
+    RefPtr<TextTemplateStmt> parseTextTemplateBody()
+    {
+        bool isAtStartOfLine = true;
+        bool isInScriptLine = false;
+        int depthInSplice = 0;
+
+        char const* currentLineBegin = _cursor;
+        char const* currentSpanBegin = _cursor;
+        while (!atEnd())
+        {
+            char const* currentSpanEnd = _cursor;
+
+            bool wasAtStartOfLine = isAtStartOfLine;
+            isAtStartOfLine = false;
+
+            int c = *_cursor++;
+            switch (c)
+            {
+            default:
+                break;
+
+            case '\r':
+                if (*_cursor == '\n')
+                {
+                    _cursor++;
+                }
+            case '\n':
+                isAtStartOfLine = true;
+                currentLineBegin = _cursor;
+                if (isInScriptLine)
+                {
+                    addScriptStmtLine(currentSpanBegin, currentSpanEnd);
+                    isInScriptLine = false;
+                    currentSpanBegin = currentSpanEnd;
+                }
+                break;
+
+            case ' ':
+            case '\t':
+                isAtStartOfLine = wasAtStartOfLine;
+                break;
+
+            case '%':
+                if (wasAtStartOfLine && !depthInSplice)
+                {
+                    addRaw(currentSpanBegin, currentLineBegin);
+                    isInScriptLine = true;
+                    currentSpanBegin = _cursor;
+                }
+                break;
+
+            case '$':
+                if (isInScriptLine)
+                    continue;
+                if (depthInSplice)
+                    throw 99;
```

Comment:
output a diagnostics instead of throwing?

Reply:
Yes, I was lazy at several points while developing this and did a quick `throw` instead of diagnosing properly.
We may still want to fail-fast on errors rather than try to soldier on, just given that the tool itself isn't 100% robust against bad input.

Reply:
you can do a fatal error style handling, just write to diagnostic and throw, and make sure we handle the exception in main and print out the diagnostics.